src/bun.js/api/bun/h2_frame_parser.zig

```diff
@@ -0,0 +1,2440 @@
+        var handler = this.handlers;
+        defer handler.unprotect();
+        this.handlers.clear();
+
+        return JSC.JSValue.jsUndefined();
+    }
+
+    const Payload = struct {
+        data: []const u8,
+        end: usize,
+    };
+
+    // Default handling for payload is buffering it
+    // for data frames we use another strategy
+    pub fn handleIncommingPayload(this: *H2FrameParser, data: []const u8, streamIdentifier: u32) ?Payload {
+        const end: usize = @min(@as(usize, @intCast(this.remainingLength)), data.len);
+        const payload = data[0..end];
+        this.remainingLength -= @intCast(end);
+        if (this.remainingLength > 0) {
+            // buffer more data
+            _ = this.readBuffer.appendSlice(payload) catch @panic("OOM");
+            return null;
+        } else if (this.remainingLength < 0) {
+            this.sendGoAway(streamIdentifier, ErrorCode.FRAME_SIZE_ERROR, "Invalid frame size", this.lastStreamID);
+            return null;
+        }
+
+        this.currentFrame = null;
+        return .{
+            .data = payload,
+            .end = end,
+        };
+    }
+
+    pub fn handleWindowUpdateFrame(this: *H2FrameParser, frame: FrameHeader, data: []const u8, stream: ?*Stream) usize {
+        // must be always 4 bytes (https://datatracker.ietf.org/doc/html/rfc7540#section-6.9)
+        if (frame.length != 4) {
+            this.sendGoAway(frame.streamIdentifier, ErrorCode.FRAME_SIZE_ERROR, "Invalid dataframe frame size", this.lastStreamID);
+            return data.len;
+        }
+
+        if (handleIncommingPayload(this, data, frame.streamIdentifier)) |content| {
+            const payload = content.data;
+            const windowSizeIncrement = UInt31WithReserved.fromBytes(payload);
+            this.readBuffer.reset();
+            // we automatically send a window update when receiving one
+            this.sendWindowUpdate(frame.streamIdentifier, windowSizeIncrement);
+            if (stream) |s| {
+                s.windowSize += windowSizeIncrement.uint31;
+            } else {
+                this.windowSize += windowSizeIncrement.uint31;
+            }
+            log("windowSizeIncrement stream {} value {}", .{ frame.streamIdentifier, windowSizeIncrement.uint31 });
+            return content.end;
+        }
+        // needs more data
+        return data.len;
+    }
+
+    pub fn decodeHeaderBlock(this: *H2FrameParser, payload: []const u8, stream_id: u32, flags: u8) void {
+        log("decodeHeaderBlock", .{});
+
+        var header_buffer: [MAX_HPACK_HEADER_SIZE]u8 = undefined;
+        var offset: usize = 0;
+
+        const globalObject = this.handlers.globalObject;
+
+        const headers = JSC.JSValue.createEmptyArray(globalObject, 0);
+
+        while (true) {
+            const header = this.decode(&header_buffer, payload[offset..]) catch break;
+            offset += header.next;
+            var result = JSValue.createEmptyObject(globalObject, 2);
```

Comment:
this is going to be very slow. 

if we are creating objects as pairs, it should not be `{name, value}`. it should be an array of tuples.

or it should be a single flat array.


if we absolutely must create pairs of `name, value`, then they need to use a single `JSC::Structure` and call `putDirectOffset` instead of creating a new structure with new key/value pairs each time

Reply:
I will transform to use the same format as nodejs events, so we would avoid processing this value twice